### PR TITLE
Enhanced apple iie

### PR DESCRIPTION
**Pull request by sethkush**
Hi, this branch adds enhanced //e support. It all seems to be working correctly as far as I can tell. I ran into issues with the alternate charset, so ended up changing the logic for character display quite a bit to fix it.

### CODE REVIEW

**Comment by sethkush** [comment]
I should note that my distro doesn't package cc65, so I haven't run those tests yet.

**Comment by sethkush** [comment]
I've now installed cc65 and fixed what was left to test. All tests pass.

**Comment by micahcowan** [comment]
Hello Seth, and thanks very much for your tremendous work on this! It clearly represents a lot of effort on your part. I'm very thankful that someone else is willing to put the work in to get 65C02 stuff working, as TBH the CPU was by far the most tedious effort in my work so far on this project, and I had slowed in my progress on this project in part due to the major task that getting 65C02 working properly would represent. I was particularly overjoyed to finally be able to fire up 4am's [Pitch Dark](https://archive.org/download/PitchDark) on bobbin!

I'm particularly grateful for the work you put in to add copious tests for your code.

I'm curious what inspired you to put this work in, as opposed to just using an existing project that already has enhanced IIe support. Are there particular features in Bobbin that you found compelling/useful for your own purposes, or was it just a codebase you found yourself wanting to work with?

Given how great an effort you've already put into this project, it pains me to say that there is still significant work ahead before I can accept this patch. I'll take this opportunity to observe that early and frequent communication with project maintainers is crucial to avoiding wasted time and energy, and that the bigger the change is, the more risky it is to forego such communication. It's likely I could have saved you some trouble, if we'd been in communication.

Before I begin discussing your changes, I'd first like to point out that I've added my own [enhanced apple-iie](https://github.com/micahcowan/bobbin/tree/enhanced-apple-iie) branch at my original repo. This consists of all your commits, plus some commits of mine that:

- reverts all changes in interfaces/*.c, mem.c and util.c (I'll explain why below, and in my code notes in your pull request).
- splits test/tests6502/6502_decimal_test.ca65 into two versions, one for 6502, and one for 65C02 (failing).
- adds `examples/all-chars.bas`, an AppleSoft program that puts all possible chars to the screen, then lets you toggle between normal and alternate charsets, to demonstrate an issue with your altchar code.
- makes a few other, minor changes.
- keeps all of your CPU code

If you want to examine my version of the branch without clobbering your own, you'll probably have to pull it into a branch with a different name. Alternatively, you may wish to clone a second copy of the repo, so you can have different working dirs simultaneously, one with your version, one with mine.

I'll begin by discussing your changes that I reverted away. I'll have more details about it in my comments on the "files changed" section of this pull request, but here's an overview.

Your display changes have broken the alt charset stuff, which was already working before you changed it. If you run the `examples/all-chars.bas` BASIC program in my version of the enhanced-apple-iie branch, and then hit keys to toggle back and forth between normal, you will get a display of all characters. Please compare this running in:
 - a different emulator (of an enhanced //e)
 - your version of bobbin
 - my branch, reverting the display changes

I gather that you were attempting to fix an issue you were seeing where border lines were getting drawn over. Unfortunately, in the process of fixing that, you added a lot of code that also:
 - fixed a number of things that weren't broken
 - broke some things that were fixed (the charset stuff above, at least)
 - introduced a new/redundant `tty_char_display()` fn whose purpose seems to be the same as `util.c`'s `util_todisplay()`
 - uses `util_todisplay()` inappropriately in `interfaces/simple.c` (in place of `util_toascii`), and then adds extra code to deal with the fact that the values returned there are not appropriate for the `simple` interface
 - considerations about what character to display to the tty, previously the sole domain of `util_todisplay()`, has now leaked out into the refresh code as well (where it maps unprintable characters to SPACE, a choice I don't support, but that code never actually triggers anyway (your `tty_char_display` function never returns characters of that kind)).
 - results in code that (to my eye, at least) is a lot less readable, with little advantage that I can find.

If there was indeed something broken, I'd rather start from scratch with a more surgical approach to the fix. I tried to run your `debug_80col_display.py` script, and also to run your added davex image by hand along the lines the script does, but I didn't see any issues. I'd appreciate an easy-to-reproduce recipe for producing the issue you were seeing, in my reversion branch, and also information about what terminal emulator and curses implementation you're using on your system.

As for `mem.c,` you seem to have changed the code there under some kind of misunderstanding: both enhanced and the unenhanced Apple //e have 16kb ROMs, and they both start at $C000, *not* $D000. If you examine the contents of those ROMs, they both contain largely nothing for the first 256 bytes, and then real contents begin. This is because the first 4 kilobytes are hidden by default, but can be swapped in, in place of peripheral ROM area ($C0000). The first page is never swapped in, as it's always hidden by the soft switches in $C0XX, which is why nothing useful is written to that first page of the ROM file. Thus, an access to $D000 should always result in an access of the `rombuf` starting at 0x1000. Accesses to `rombuf` lower than that are handled by `slot_access_area_sneaky` (also in `mem.c`). The fact that `mem_get_true_access` doesn't handle the case of ROM accesses from $CXXX is documented by a comment in `bobbin-internal.h` (though perhaps a duplicate comment should be included in `mem.c`). Ultimately, because you wound up adding 0x1000 to your resulting value, your math wound up the same as my original, but I feel the original was less confusing.

All of the above, of course, is a lot of explanation over something pretty easily fixed: I just reverted all of it. It is, of course, possible that I may have thrown out some genuinely useful or important changes along with the unhelpful ones, but it's too hard to tell, especially without a test I can use to tell me what I'd been doing wrong. I'm happy to restart on a solution to whatever you felt needed fixing, but I do believe we'll have to start from square one there, as I can't accept this version of these changes.

**Comment by micahcowan** [comment]
Now let's talk about the CPU changes.

When illegal/undocumented 6502 opcodes were run on my previous code, it would trigger the same behavior as `BRK` (except for a couple of opcodes I found it necessary to handle specially, as they're used by some important software to distinguish the CPU type (6502 vs 65C02 vs 65816), based on checking the behavior of those opcodes.

With your changes, the behavior of those opcodes has changed to be that it does whatever the next-supported 6502 opcode that comes numerically (or, at least, the next `case` label) does, because of the `/* FALLTHROUGH */` behavior for those codes. I'm not really happy with this turn of events. Ideally, they shouldn't do what they did before, either - they should behave exactly as the real hardware does. But I haven't found a lot of great information for precisely emulating all undocumented opcodes, so in the meantime, behaving like `BRK` and hitting either the firmware monitor, or else **bobbin**'s debugger, is a reasonable replacement.

I very much appreciate you including a suite of tests for the new 65C02 opcodes, and fixing issues that arose from them. They do, however, fail to behave correctly for decimal arithmetic. Specifically, they currently behave exactly like a 6502, which is not correct. Ideally I'd like to address that before including this code in a release, but I might just have to note it as a bug to be fixed at some future date. I'm not sure, but I think there may also be some non-decimal differences in behavior for `SBC` as well, can't remember. My `enhanced-apple-iie` branch has added tests for 65C02 decimal arithmetic, to target against. But debugging that stuff tends to be a real PITA (part of why I hadn't gotten to this work). I would imagine that the current version of your CPU code behaves correctly when given legal operands, and perhaps some improper flag values even for legal operands. A lot of opcode test suites don't bother with those things (which is why I made a point to include a separate set of tests specifically for those)

I haven't looked closely at the 65C02 tests you included, but assuming they cover everything else, I can assume that the actual behavior of your CPU code, in terms of input and output, is correct (apart from BCD and maybe some other corner cases). However, it's pretty certain that the *timing* of a number of operations, in terms of cycles, will not be right, as those differ for a number of opcodes between 6502 and 65C02, and I'm not seeing any updates for those. Presumably the behavior, at the level of what kinds of bus accesses happen when, may also be incorrect. At this stage in bobbin's development, that has basically no actual repercussions. Bobbin has no graphics support, and even more importantly, no sound support. But these features are planned, and for both of those things, as well as possible future support of obscure peripherals, I'm going to need to be accurate with timings, where possible. Some peripherals also depend heavily on specific timings for bus actions, as I'm given to understand. It's my intention to eventually support exotic graphics techniques that "race the beam", in order to achieve graphical effects otherwise impossible on the 8-bit Apple computers, and I'm trying to have an eye toward those goals, with the groundwork I lay today.

**Comment by micahcowan** [comment]
Aside from what's discussed above, there's been some files under `test/pexpect` that were added, but later removed from being run. If those tests are no longer applicable, then the files should be removed; but it wasn't yet clear to me whether you still had purposes for some of those. If there are reasons to keep some defunct tests around but not run them by default during `make check`, then we should document why they're there. I did remove (from my branch) what appeared to be a number of accidentally-checked in artifacts from running some tests by hand (not via `make check`), under `test/noninteractive`. I've removed them, but if this is eventually going in the `main` branch, then it'd be best if there are no commits including them in the first place. I'd also like to omit the DAVEX ProDOS shell hard drive image from any final inclusion. If it's still necessary to exercise some bug test, then it should stay for now, but we should work together to find/create a smaller replacement that exercises the same problem, without obliging us to include this image.

The rough roadmap for getting this branch into a happier state would be (not all of this has to be on you):
- Pruning any remaining, extraneous files that aren't needed/being used. Removal of accidentally-included files, and hdd images we won't use moving forward, should eventually be removed at the commit history level.
- Restore my original code logic for `interfaces/*`, `util.c` and `mem.c` (I've already done this in my version of the branch, but may want to remove it at the history level as well, before inclusion; it would make for a cleaner history). Any *actually* necessary changes can be added back in again, anew.
- Fix “fallthrough” of 65C02 operations to the next supported opcode: illegal ops in 6502 should behave as before (same as BRK), or else like they do in the real CPU.
- Update the `README.md`
- Fix behavior of BCD on 65C02 (*may not* be strictly necessary before inclusion, I have to consider it further)
- Fix various operation timings/cycle-by-cycle behavior on 65C02 (*may not* be strictly necessary before inclusion, I have to consider it further) (ideally I should get tests in place for this stuff, shouldn't assume your 65C02 stuff *or* my 6502 stuff is valid until I have a way to verify it...)
- Fix `make distcheck` so it works again. This especially happens when new files are added, but `automake` isn't made aware of them (in `Makefile.am` files) in a way that makes it realize it should include them when packaging up tarball. It may even have already been broken before your changes but it's definitely been broken further, and that needs to be fixed before I could release with it. I made some untested changes relevant to this in `test/tests6502/Makefile.am`, but I think more elsewhere are needed. 
- ProDOS 2.0.3 fails to detect that it's running under an enhanced 65C02, that'll need fixing
- Add 65C02 support to **bobbin**"s built-in disassembly facilities (that probably has to be me I guess; doesn't need to hold up merging, either, just throwing it in here).

File: disk/davexuiowa/uiowa127.hdv
Comment (micahcowan) [inline disk/davexuiowa/uiowa127.hdv:1]: Don't think I want to include this in the distribution. If it's needed to exercise a test, we need something more basic if possible.

File: src/cpu.c [hunk 1]
```diff
@@ -447,6 +447,15 @@ void cpu_step(void)
...
             break;
+        case 0x04: // TSB zp - Test and Set Bits, Zero Page (MOS 65C02 only)
+            if (machine_is_enhanced_iie()) {
+                OP_RMW_ZP(
+                    PPUT(PZERO, (val & ACC) == 0);
+                    val |= ACC;
+                );
+                break;
+            }
+            /* FALLTHROUGH */
```
Comment (micahcowan) [inline src/cpu.c:458]: This fallthrough makes opcode $04 behave like it was $05, rather than defaulting to the `BRK` handling as it did previously.

File: src/interfaces/simple.c [hunk 2]
```diff
@@ -480,7 +480,10 @@ void vidout(void)
         // (emit, don't emit), we must stop suppressing.
         output_suppressed = SUPPRESS_NONE;
     }
-    int c = util_toascii(ACC);
```
Comment (micahcowan) [inline src/interfaces/simple.c:483]: This change is inappropriate. Making this change, to use a function ill-suited to this purpose, forced you to "fix" `util_todisplay` so it no longer returns visible ASCII characters for control characters. But it needs to do that so that "control" chars on the screen are displayed properly in tty mode. So you created a replacement function in `tty.c` to do what `util_todisplay` used to... except your replacement wound up being somewhat busted, and I don't even know what `util_todisplay` is supposed to do now.

File: src/interfaces/tty.c [hunk 3]
```diff
@@ -94,15 +95,15 @@ static void draw_border(void)
 {
     int y, x;
     getmaxyx(stdscr, y, x);
-    if (x > cols) {
+    if (x > cols + 2) {  // Need at least 2 extra columns for border + safety margin
```
Comment (micahcowan) [inline src/interfaces/tty.c:98]: We don't need a horizontal safety margin. It's perfectly valid for the cursor to go to the next line if we're using a terminal emulator that does that, because we don't draw any further characters after, without first repositioning again.

And if we did need one, we'd only need a `+ 1` here, because we're using `>` and not `>=`.

File: src/interfaces/tty.c [hunk 4]
```diff
@@ -94,15 +95,15 @@ static void draw_border(void)
...
         move(0,cols);
         vline('|', y >= 25? 25: y);
     }
     if (y > 24) {
         move(24,0);
         hline('-', x >= (cols)? (cols): x);
     }
-    if (x > cols && y > 24) {
+    if (x > cols && y > 24 && cols < x) {
```
Comment (micahcowan) [inline src/interfaces/tty.c:106]: `x > cols && cols < x`? I'm not seeing how this helps.

File: src/interfaces/tty.c [hunk 5]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
         for (int x=0; x != 40; ++x) {
             byte c = mem[base + x];
             if (util_isflashing(c)) {
-                byte cd = util_todisplay(c);
+                byte cd = tty_char_display(c);
```
Comment (micahcowan) [inline src/interfaces/tty.c:143]: We don't need a new function, this is exactly what `util_todisplay`'s job is. Change that, if necessary.

File: src/interfaces/tty.c [hunk 6]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
...
     attrset(A_NORMAL);
 }
 
+
 static void refresh_video80(void)
 {
     attrset(A_NORMAL);
     const byte *membuf = getram();
     bool have_aux = cfg.amt_ram > LOC_AUX_START;
+    bool flash = (cols == 80 || swget(ss, ss_altcharset))? false : saved_flash;
```
Comment (micahcowan) [inline src/interfaces/tty.c:157]: there are no flashing characters in 80-column mode. And cols is always 80 if we're in refresh_video80, so `flash` is always set to `false`. Which is what we were already passing to `util_isreversed`.

File: src/interfaces/tty.c [hunk 7]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
...
-        for (byte x=0; x != 80; ++x) {
-            bool even = have_aux && (x % 2 == 0);
+        // Account for border and cursor advancement - if we're drawing a border, leave space for it
+        // Also, avoid writing to the last column to prevent cursor advancement issues
+        int max_x = 80;
+        if (COLS > cols + 1) {
+            max_x = cols; // Border will be at 'cols', so content stops before that
+        } else {
+            max_x = COLS - 1; // Leave last column to prevent cursor advancement past screen
+        }
```
Comment (micahcowan) [inline src/interfaces/tty.c:169]: We don't need or want the above checks. It's the job of `refresh_video` to first ensure there's enough room to draw all of the content of the screen, and complain if there's not; so we don't need to limit `x`'s bounds, we already know it has space for that. There shouldn't be any consideration given to borders at all here: it's `draw_border()`'s job to decide to draw it, if and only if there's room to do so. We always draw the full contents of the screen, I don't want to clip any of it at all (and if we were going to clip it, we should have strong visual signalling to that effect).

We also don't need to leave last columns, because again, we always explicitly set location after we've written the last character.

File: src/interfaces/tty.c [hunk 8]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
...
+            
+            if (have_aux && x < 80) {
+                // True 80-column mode with auxiliary memory
+                bool even = (x % 2 == 0);
+                c = membuf[(base | (even? LOC_AUX_START : 0)) + mx];
+            } else {
+                // 40-column mode or fallback - use main memory only
+                mx = x < 40 ? x : 39;  // Prevent reading past buffer
+                c = membuf[base + mx];
+            }
```
Comment (micahcowan) [inline src/interfaces/tty.c:182]: This check is redundant. We already do it in `if_tty_poke`, where if we detect main video memory being written to, we copy that same video memory to where aux mem *would've been*, so it gets picked up here. (Granted, perhaps a comment should document that here)

File: src/interfaces/tty.c [hunk 9]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
...
+            byte cd = tty_char_display(c);
+            
+            bool cfl = util_isreversed(c, flash);
+            
+            // Handle null characters and other control characters for terminal display
+            if (cd == 0x00) {
+                cd = ' '; // Convert null to space for terminal display
+            } else if (cd < 0x20 || cd > 0x7E) {
+                cd = ' '; // Convert other non-printable characters to space
+            }
```
Comment (micahcowan) [inline src/interfaces/tty.c:193]: Deciding what char to draw is `util_todisplay()`'s job. Its pervue shouldn't leak out into this function. Also we don't want spaces instead of control characters or non-printable characters: all screen characters are printable, so everything returned by `util_todisplay()` or `tty_char_display()` should be printable characters. If they're not, we should abort the program and ask to send a bug report, rather than hide it by emitting spaces.

File: src/interfaces/tty.c [hunk 10]
```diff
@@ -138,29 +140,60 @@ static void repaint_flash(bool flash)
...
+            
+            // Use mvaddch to explicitly position the character and avoid cursor advancement issues
+            mvaddch(y, x, cd | (cfl? A_REVERSE: 0));
```
Comment (micahcowan) [inline src/interfaces/tty.c:196]: Shouldn't be necessary to explicitly position each character, when we're explicitly positioning after the end of every line.

File: src/interfaces/tty.c [hunk 11]
```diff
@@ -348,6 +391,12 @@ static void if_tty_switch(void)
 
     refresh_all = refresh_all || (prev_page != text_page || prevcols != cols
         || oldcharset != altcharset);
+    
+    // If switching column modes, immediately clear screen to remove garbage
+    if (prevcols != cols) {
+        clear();
+        refresh_all = true;
+    }
```
Comment (micahcowan) [inline src/interfaces/tty.c:399]: This is covered by the statement immediately preceding it. It doesn't call clear(), but things seem to get cleared due to refresh_all being true (if that is incorrect, then we should fix it: but it should happen by putting all of the above into the `if`'s condition, and not just this).

File: src/interfaces/tty.c [hunk 12]
```diff
@@ -634,6 +697,78 @@ static void if_tty_event(Event *e)
     }
 }
 
+// TTY-specific character display based on proper Apple IIe technical documentation
+static byte tty_char_display(byte c)
+{
```
Comment (micahcowan) [inline src/interfaces/tty.c:702]: As mentioned elsewhere, this function is trying to do the job of a function that already exists (`util_todisplay()`, before your modifications to it), and does it more poorly (incorrect mapping of characters in the range of 0x60 to 0x7F, which should be lowercase when altchar, but should be flashing punctuation and digits normally).

File: src/mem.c [hunk 13]
```diff
@@ -344,7 +344,9 @@ void mem_get_true_access(word loc, bool wr, size_t *bufloc, bool *in_aux, MemAcc
                          : !swget(ss, ss_lc_read_bsr)))) {
         *access = MA_ROM;
         size_t romsz = expected_rom_size();
-        *bufloc = loc - (LOC_ADDRESSABLE_END - romsz);
+        // For Enhanced Apple IIe (16KB ROM), use offset to map ROM starting at 0xD000
+        size_t rom_offset = (romsz == 16 * 1024) ? 0x1000 : 0;
+        *bufloc = loc - LOC_ROM_START + rom_offset;
```
Comment (micahcowan) [inline src/mem.c:349]: Same math results as the original, but less clear (to me, anyway) about what and why it's doing. The comment in particular is either wrong, or else confusing. (Both enhanced and unenhanced IIe ROMs map in memory starting at $C000 (but $C000.C0FF is never available to read). $C000.CFFF is handled in a different function; if we got here we're at least in $D000, which means we're at offset 0x1000 or higher to the rom buffer.... but the original way of calculating that is simpler and therefore (IMO) clearer.

File: src/util.c [hunk 14]
```diff
@@ -94,33 +94,43 @@ bool util_isreversed(int c, bool flash)
 int util_todisplay(int c)
 {
     if (c < 0) return c;
+    
+    // Handle control characters first - preserve ALL non-printing control characters
+    // This applies to both Apple II+ and Apple IIe machines
+    int control_char = c & 0x7F;
+    if (control_char < 0x20) {
+        return control_char;
+    }
```
Comment (micahcowan) [inline src/util.c:103]: If we're going to make a function that calls itself `util_todisplay()` return non-printable characters, it seems to me we should call it something else. It's not at all clear to me what it's supposed to be now: it used to serve printable characters for tty-mode, but now that uses something else. Now it serves characters to be used by simple mode, but that already had `util_toascii`, which did a better job.

File: test/noninteract/run_basic.t/reverse
Comment (micahcowan) [inline test/noninteract/run_basic.t/reverse:1]: accidentally included

File: test/noninteract/disk_do_rw.t/testdisk.dsk
Comment (micahcowan) [inline test/noninteract/disk_do_rw.t/testdisk.dsk:1]: accidentally included

File: test/noninteract/disk_two_prodos.t/testdisk-a.dsk
Comment (micahcowan) [inline test/noninteract/disk_two_prodos.t/testdisk-a.dsk:1]: accidentally included

File: test/noninteract/disk_two_prodos.t/testdisk-b.dsk
Comment (micahcowan) [inline test/noninteract/disk_two_prodos.t/testdisk-b.dsk:1]: accidentally included

File: test/pexpect/bcd_detection.py
Comment (micahcowan) [inline test/pexpect/bcd_detection.py:1]: this file is no longer included in test runs. Should it be deleted?

File: test/pexpect/charset.py
Comment (micahcowan) [inline test/pexpect/charset.py:1]: this file is no longer included in test runs. Should it be deleted?

File: test/pexpect/cpu65c02.py
Comment (micahcowan) [inline test/pexpect/cpu65c02.py:1]: this file is no longer included in test runs. Should it be deleted?

File: test/pexpect/cpu_verification.py
Comment (micahcowan) [inline test/pexpect/cpu_verification.py:1]: this file is no longer included in test runs. Should it be deleted?

File: test/pexpect/debug_80col_display.py
Comment (micahcowan) [inline test/pexpect/debug_80col_display.py:1]: this file is no longer included in test runs. Should it be deleted?
It's also unclear what it's trying to get DAVEX to do. When I type those things in the booted hdv image you checked in, it doesn't do useful things. Did perhaps the .hdv image get changed in some incompatible way since you wrote this test?

File: test/pexpect/enhanced_detection.py
Comment (micahcowan) [inline test/pexpect/enhanced_detection.py:1]: this was removed from run_tests.py. Should it be deleted?

File: test/pexpect/mousetext.py
Comment (micahcowan) [inline test/pexpect/mousetext.py:1]: this was removed from run_tests.py. Should it be deleted?
Holdover from when you were distinguishing different mousetext chars, before you standardized (back) to '@'.

File: test/pexpect/tty_tests.py
Comment (micahcowan) [inline test/pexpect/tty_tests.py:1]: this was removed from run_tests.py. Should it be deleted?

File: test/tests6502/Makefile.am [hunk 15]
```diff
@@ -3,23 +3,39 @@ SRCS = \
 	6502_decimal_test.ca65 \
 	6502_functional_test.ca65
 
+# 65C02 tests - only run on Enhanced Apple IIe
+SRCS_65C02 = \
+	65c02_stack_test.ca65 \
+	65c02_bra_test.ca65 \
+	65c02_stz_test.ca65 \
+	enhanced_iie_test.ca65
+
```
Comment (micahcowan) [inline test/tests6502/Makefile.am:12]: Since this var's contents don't get included in `$(SRCS)`, the files associated with it don't get included in the distribution tarball. I fixed this (untested) in my version of the branch. Hopefully.

Well, I mean you put `$(SRCS)` and `$(SRCS_65C02)` in `EXTRA_DIST`, so I guess we're fine either way. I don't remember why I had `$(SRCS)` in there, since that should be automatic. Maybe there were conditionally0generated source files at the time, or maybe there's some logic about it not getting included because ca65 wasn't found and that messed with DIST somehow, idk.

**Comment by sethkush** [comment]
Hi there Micah,

Thanks for the extensive review. The answer to the "why" question is that I like bobbin's potential for automation, and was hoping to use it for testing in my development of apple ii software.

I really appreciate you taking a detailed look at the mess I've made haha. It will take me some time, but I will try to address all your points to make it mergeable in the future.

**Comment by micahcowan** [comment]
I think I overstated how much it would actually take to merge it, TBH. I said there's a lot of work to get to that point because (a) my head was full of all the work that needs doing for it, and (b) I wasn't operating on a ton of sleep. But the truth is, if you drop it and do nothing further, I think you've gotten me far enough that I'd still ultimately use what you've given me (and do the roadmap items myself). Even if I do absolutely nothing further to it (apart from a bit of cleanup, like the no-longer-used test files, and the switch fall-throughs in `cpu.c`, etc), merging it in its current state (with the interface code rolled back) gives me working //e support (modulo ProDOS's failed detection) where there wasn't any before, and honestly it's not until **bobbin** has that support that I can consider it a "proper Apple II emulator".

If you'd like to continue working on the things I've mentioned, I'll certainly be grateful for it (and if you do, let me know what you'll work on, so I'm not duplicating the effort myself), but the only thing I *truly* need from you, is to know the specifics of what issue you were experiencing in tty mode that you needed to fix, and how I can reliably reproduce it in my current version, to make sure it's fixed (especially if I've now thrown out your fixes for it). Also, if you have responses to my complaints about your changes in the interface code, I do want to hear them - I'm not trying to be some kind of prima donna about my old code, I just want changes to be clearly justified, and as readable/understandable as feasible.

Thanks again for your tremendous work on this! Despite my whining about some aspects of it, it truly is very much appreciated, and I can see you spent a lot of dedicated work on it. And of course I'm always tickled when I find someone besides me who also thinks this "yet another" Apple II emulator actually *was* a good idea.

(Oh yeah - I forgot the documentation will also need to be updated to account for your changes. I'll add that to my roadmap.)

**Comment by micahcowan** [comment]
I've created an [enhanced-revised](https://github.com/micahcowan/bobbin/tree/enhanced-revised) branch, which is more-or-less the same as my edition of the `enhanced-apple-iie` branch, except:
- Instead of removing things via later commits on my part, I edited your commits to omit them in the first place (and annotated the commit log to explain that). *(So, this branch shares no history with the `enhanced-apple-iie` branch (yours nor mine).)*
- I removed the DAVEX stuff, and the tests that you later disconnected from `run_tests.py`
- I removed trailing whitespace from lines in some of the newly-added files
- I added your name to the `./THANKS` file. :slightly_smiling_face: 

I intend to use this revised branch as the foundation for further work on enhanced //e stuff

**Comment by sethkush** [comment]
Hey again,

I think best would be to let you take over doing the integration your way. I was running into many problems with mousetext display in the tty mode, likely because I don't fully understand what's going on there. At one point I seemed to be triggering a bug in curses where junk I couldn't account for was being output to the right of the screen; that's when I hackily reworked the tty stuff.

If you don't mind, I'll focus my effort on CPU accuracy. I think I can make good contributions to that.

I'm not sure how you'd feel about merging a 65802 option, but as I have one in my //e, that would be something I'd be very interested in emulating down the road.

**Comment by micahcowan** [comment]
Sounds good to me. Before you get started on more CPU stuff: I think I'm going to split out all the new 65C02 stuff into its own separate function from the common-to-6502-and-65C02 stuff. Ideally at some future point I want to make illegal opcodes on 6502 more accurate, and I feel at that point the CPU code will become significantly harder to follow with all the if/else statements, without separating the CPUs a little more. I'll try and get to that today (in the `enhanced-revised` branch), and then you can use that as a base for the accuracy improvements.

Re: 65802, yes, I believe I would merge that, provided that they have verification tests and it doesn't appear that it would be a significant maintenance burden (I don't see why it would be). I'd treat it as an "experimental" feature, and reserve the right to remove it in the future if it started to become a hassle (again, don't see why that would be so, especially if there are handy tests I can always run against).

**Comment by micahcowan** [comment]
Re: the moustext stuff and the ncurses bug/junk being printed to the right of the screen: can you help me test whether you're still experiencing that, in `enhanced-revised`? I want to be sure that we do in fact fix anything like that you were seeing, whether that means pulling back in some portions of your changes, or (me) reworking a fix. If it's possible that the bug itself was only introduced when you were working on mousetext, then there's probably nothing to fix, but if your issue was pre-existing/hiding and you exposed it, then it needs to be fixed.

When I fire up your DAVEX hdv image, I just get a command prompt without any interesting mousetext stuff. I don't really know what I'm doing in there, but the 80-column debug stuff you had in `test/pexpect` seemed to expect a different environment than I was seeing when I fire it up myself, so my thought was that perhaps the .hdv was in a different state than the expected one for that test.

**Comment by micahcowan** [comment]
I'm going to close the Pull Request here, since I'm not planning to merge this version of the branch, but created the revised version I do plan to merge at some point.

**Comment by micahcowan** [comment]
Updating this PR to note that the revised version was indeed merged, in commit 0665fb7a57ae8e3935278d9c1c16d3826395f51b. (No merge commit: it was a rebase-and-reset). Thanks again for your work!
